cnd/src/swap_protocols/hbit.rs

```diff
@@ -0,0 +1,204 @@
+async fn create_watcher(
+    connector: Arc<Cache<BitcoindConnector>>,
+    ledger_states: Arc<LedgerStates>,
+    swap_id: LocalSwapId,
+    htlc_params: HtlcParams,
+    accepted_at: NaiveDateTime,
+) {
+    ledger_states
+        .insert(swap_id, LedgerState::NotDeployed)
+        .await;
+
+    // construct a generator that watches alpha and beta ledger concurrently
+    let mut generator =
+        Gen::new({ |co| async { watch_ledger(connector, co, htlc_params, accepted_at).await } });
+
+    loop {
+        // wait for events to be emitted as the generator executes
+        match generator.async_resume().await {
+            // every event that is yielded is passed on
+            GeneratorState::Yielded(event) => {
+                tracing::info!("swap {} yielded event {}", swap_id, event);
+                ledger_states.update(&swap_id, event).await;
+            }
+            // the generator stopped executing, this means there are no more events that can be
+            // watched.
+            GeneratorState::Complete(Ok(_)) => {
+                tracing::info!("swap {} finished", swap_id);
+                return;
+            }
+            GeneratorState::Complete(Err(e)) => {
+                tracing::error!("swap {} failed with {:?}", swap_id, e);
+                return;
+            }
+        }
+    }
+}
+
+/// Returns a future that waits for events to happen on a ledger.
+///
+/// Each event is yielded through the controller handle (co) of the coroutine.
+async fn watch_ledger(
+    connector: Arc<Cache<BitcoindConnector>>,
+    co: Co<SwapEvent>,
+    htlc_params: HtlcParams,
+    start_of_swap: NaiveDateTime,
+) -> anyhow::Result<()> {
+    let deployed = connector.htlc_deployed(&htlc_params, start_of_swap).await?;
+    co.yield_(SwapEvent::Deployed(deployed.clone())).await;
+
+    let funded = connector
+        .htlc_funded(&htlc_params, &deployed, start_of_swap)
+        .await?;
+    co.yield_(SwapEvent::Funded(funded)).await;
+
+    let redeemed = connector.htlc_redeemed(&htlc_params, &deployed, start_of_swap);
+
+    let refunded = connector.htlc_refunded(&htlc_params, &deployed, start_of_swap);
+
+    match future::try_select(redeemed, refunded).await {
+        Ok(Either::Left((redeemed, _))) => {
+            co.yield_(SwapEvent::Redeemed(redeemed.clone())).await;
+        }
+        Ok(Either::Right((refunded, _))) => {
+            co.yield_(SwapEvent::Refunded(refunded.clone())).await;
+        }
+        Err(either) => {
+            let (error, _other_future) = either.factor_first();
+
+            return Err(error);
+        }
+    }
+
+    Ok(())
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct HtlcParams {
+    pub network: bitcoin::Network,
+    pub asset: asset::Bitcoin,
+    pub redeem_identity: identity::Bitcoin,
+    pub refund_identity: identity::Bitcoin,
+    pub expiry: Timestamp,
+    pub secret_hash: SecretHash,
+}
+
+// TODO: Remove the Herc20 stuff i.e., no separate Deployed/Funded states.
```

Comment:
This should be tracked somewhere else. 
I propose to add a comment above at around line 135 explaining why we have this right now and creating a ticket to remove the unneeded swap events.